ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java

```diff
@@ -1503,24 +1503,31 @@ public static Map<Path, HdfsDirSnapshot> getHdfsDirSnapshotsForCleaner(final Fil
     stack.push(fs.listStatusIterator(path));
     while (!stack.isEmpty()) {
       RemoteIterator<FileStatus> itr = stack.pop();
-      while (itr.hasNext()) {
-        FileStatus fStatus = itr.next();
-        Path fPath = fStatus.getPath();
-        if (acidHiddenFileFilter.accept(fPath)) {
-          if (baseFileFilter.accept(fPath) ||
-                  deltaFileFilter.accept(fPath) ||
-                  deleteEventDeltaDirFilter.accept(fPath)) {
-            addToSnapshoot(dirToSnapshots, fPath);
-          } else {
-            if (fStatus.isDirectory()) {
-              stack.push(fs.listStatusIterator(fPath));
+      try {
+        while (itr.hasNext()) {
+          FileStatus fStatus = itr.next();
+          Path fPath = fStatus.getPath();
+          if (acidHiddenFileFilter.accept(fPath)) {
+            if (baseFileFilter.accept(fPath) ||
+                    deltaFileFilter.accept(fPath) ||
+                    deleteEventDeltaDirFilter.accept(fPath)) {
+              addToSnapshoot(dirToSnapshots, fPath);
             } else {
-              // Found an original file
-              HdfsDirSnapshot hdfsDirSnapshot = addToSnapshoot(dirToSnapshots, fPath.getParent());
-              hdfsDirSnapshot.addFile(fStatus);
+              if (fStatus.isDirectory()) {
+                stack.push(fs.listStatusIterator(fPath));
+              } else {
+                // Found an original file
+                HdfsDirSnapshot hdfsDirSnapshot = addToSnapshoot(dirToSnapshots, fPath.getParent());
+                hdfsDirSnapshot.addFile(fStatus);
+              }
             }
           }
         }
+      }catch(FileNotFoundException fne){
+        //Ignore....
+        //As current FS API doesn't provide the ability to supply a PathFilter to ignore the staging dirs,
+        // need to catch this exception
```

Comment:
There is something like RemoteIterators.filteringRemoteIterator(fs.listStatusIterator(path), <filter Condition>) provided by Hadoop for such use cases.

Reply:
Thanks Ayush for this. I have made necessary changes with RemoteIterators.filteringRemoteIterator